Lib9c/Action/MigrationLegacyShop.cs

```diff
@@ -0,0 +1,83 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Bencodex.Types;
+using Libplanet;
+using Libplanet.Action;
+using Nekoyume.Model.State;
+using Serilog;
+
+namespace Nekoyume.Action
+{
+    [Serializable]
+    [ActionType("migration_legacy_shop")]
+    public class MigrationLegacyShop : GameAction
+    {
+        public List<Address> SellerAvatarAddresses;
+        public override IAccountStateDelta Execute(IActionContext context)
+        {
+            var states = context.PreviousStates;
+            if (context.Rehearsal)
+            {
+                states = SellerAvatarAddresses.Aggregate(states,
+                    (current, sellerAvatarAddress) => current.SetState(sellerAvatarAddress, MarkChanged));
+
+                return states.SetState(Addresses.Shop, MarkChanged);
+            }
+
+            CheckPermission(context);
+
+            Log.Debug("Start Migration Legacy Shop");
+            var shopState = states.GetShopState();
+            var groupBy = shopState.Products.Values
+                .Where(s => SellerAvatarAddresses.Contains(s.SellerAvatarAddress))
+                .GroupBy(p => p.SellerAvatarAddress)
+                .ToList();
+
+            Log.Debug($"Found {groupBy.Count} Seller Avatar.");
+            foreach (var group in groupBy)
+            {
+                var avatarAddress = group.Key;
+                var avatarState = states.GetAvatarState(avatarAddress);
+                Log.Debug($"Start Migration Avatar({avatarAddress}). Target Count: {group.Count()}");
+                foreach (var shopItem in group)
+                {
+                    var sellerAgentAddress = shopItem.SellerAgentAddress;
+                    var agentAddress = avatarState.agentAddress;
+                    if (!sellerAgentAddress.Equals(agentAddress))
+                    {
+                        Log.Debug($"Skip Invalid ShopItem. Expected: {sellerAgentAddress}, Actual: {agentAddress}");
+                        continue;
+                    }
+
+                    if (!(shopItem.ItemUsable is null))
+                    {
+                        avatarState.inventory.AddItem(shopItem.ItemUsable);
+                    }
+                    if (!(shopItem.Costume is null))
+                    {
+                        avatarState.inventory.AddItem(shopItem.Costume);
+                    }
+                    shopState.Unregister(shopItem);
```

Comment:
How about unregister `shopItem` from `shopState` only when succeed to migration like below.
```suggestion
                    var migrated = false;
                    if (!(shopItem.ItemUsable is null))
                    {
                        avatarState.inventory.AddItem(shopItem.ItemUsable);
                        migrated = true;
                    }
                    if (!(shopItem.Costume is null))
                    {
                        avatarState.inventory.AddItem(shopItem.Costume);
                        migrated = true;
                    }
                    
                    if (migrated)
                    {
                        shopState.Unregister(shopItem);
                    }
```

Reply:
Can `AddItem` fail?

Reply:
I think the `shopItem` should remained in `shopState` when `shopItem.ItemUsable` and `shopItem.Costume` are `null`.

Reply:
If both is null, it can't register shop. so I think this case can be ignored.

Reply:
Your words are right. I intended not to even consider the registration logic of `ShopState` within the context of this action.
I wanted the `ShopItem` registered in `ShopState` to not have to know that either `ItemUsable` or `Consumable` must not be `null` when determining whether the `MigrationLegacyShop` action is appropriate.